File: opcua-webservice-spring/examples/uawebservicetest/src/main/java/org/opcfoundation/uawebservicetest/testdigitaltwin/DigitalTwinDirectoryTestType.java
package org.opcfoundation.uawebservicetest.testdigitaltwin;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.digitaltwin.DigitalTwinRepositoryType;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDigitalTwinListRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDigitalTwinListResponse;
import org.opcfoundation.webserver.types.digitaltwin.DigitalTwinDescriptor;

import java.util.concurrent.CompletableFuture;

public class DigitalTwinDirectoryTestType extends DigitalTwinRepositoryType {
    public DigitalTwinDirectoryTestType(DigitalTwinSpace space) {
        super("TestDigitalTwinDirectory", new LocalizedText("TestDigitalTwinDirectory"), space);
        setDescription(new LocalizedText("TestDigitalTwinDirectory"));
    }

    @Override
    public CompletableFuture<GetDigitalTwinListResponse> onGetDigitalTwinList(GetDigitalTwinListRequest request)
    {
        GetDigitalTwinListResponse response = new GetDigitalTwinListResponse();

        for (int i = request.getOffset(); i<request.getOffset()+5; ++i)
        {
            response.add(new DigitalTwinDescriptor(
                    Integer.toString(i),
                    new LocalizedText("DT " + i),
                    DigitalTwinSpaceTest.digitalTwinTestType));
        }

        response.setContainsMoreData(0 == request.getOffset());

        return CompletableFuture.completedFuture(response);
    }
}

File: opcua-webservice-node/examples/testserver/testdigitaltwin/DigitalTwinDirectoryTestType.ts
import { UaLocalizedText } from "opcua-webapi-ts";
import { DigitalTwinSpace } from "../../../src/server/digitaltwin";
import { DigitalTwinRepositoryType } from "../../../src/server/digitaltwin/digitaltwin/DigitalTwinRepositoryType";
import { GetDigitalTwinListRequest, GetDigitalTwinListResponse } from "../../../src/server/service/message";
import { DigitalTwinDescriptor } from "../../../src/server/types/digitaltwin/DigitalTwinDescriptor";
import { DigitalTwinSpaceTest } from "./DigitalTwinSpaceTest";

export class DigitalTwinDirectoryTestType extends DigitalTwinRepositoryType {
    constructor(space: DigitalTwinSpace) {
        super("TestDigitalTwinDirectory", new UaLocalizedText("TestDigitalTwinDirectory"), space);
        this.description = new UaLocalizedText("TestDigitalTwinDirectory");
    }

    override async onGetDigitalTwinList(request: GetDigitalTwinListRequest): Promise<GetDigitalTwinListResponse> {
        const response = new GetDigitalTwinListResponse();

        for (let index = request.offset; index < request.offset + 5; ++index) {
            response.add(
                new DigitalTwinDescriptor(
                    `${index}`,
                    new UaLocalizedText(`DT ${index}`),
                    DigitalTwinSpaceTest.digitalTwinTestType,
                ),
            );
        }

        response.containsMoreData = request.offset === 0;
        return response;
    }
}